### DISCUSSION TITLE
ClientAPI does not allow changing setting for Data Pruner

### DISCUSSION THREAD

**Discussion by mcswaip1** [nextgenhealthcare/connect#5905] [Q&A] [answered]
I was wondering if the Data Pruner settings were purposefully left out of the ClientAPI (Swagger) and if so, what was the reasoning behind the decision.
Thank you!

**Comment by jonbartels** [comment, marked as answer]
The Data Pruner is an extension (plugin) for Mirth Connect. Many of the setting panels and all of the connectors are really plugins so there isn't a first-class API for them. Its a pluggable API.

Look under "Extension Services" and you'll see `/extensions/datapruner` for operations to start/stop the pruner.

The settings are a bit more obscure, Look inder "Extensions". I think how those are set is under `/extensions/{extensionName}/properties`. Each plugin will maintain a set of properties for that plugin. These are the settings you see in the server settings for the pruner. 

Finally the pruner stores settings as channel metadata. I have a hard time explaining this but I'll try. [Look at this query on line 78-81](https://gist.github.com/MichaelLeeHobbs/67980d165fc68880eb2ab283c673244b), see how it looks up "ChannelMetadata" objects? Mirth stores things like per-channel pruner settings, tags, etc. as "ChannelMetadata". This is NOT the same as "Metadata Columns". It is one big XML blob with a channel ID then settings for the pruner and the archiver for each channel. This is in the "Server" API section under `/server/channelMetadata` . You'll note that the ENTIRE metadata blob is updated as a complete set and not individually.

There is a bit of tech debt in how Mirth handles pruner settings for each channel. I hope I have explained it well. I agree with you, this could be documented more clearly. It is available via API but it requires some deeper understanding of how Mirth works and an API is supposed to abstract all that away from the API user!

**Reply by mcswaip1** [reply]
Excellent answer, however I was really referring to the `Settings | Data Pruner`.
Your answer did point me in a direction to be able to set these values. The CONFIGURATION table has several rows with CATEGORY = 'Data Pruner'.  I could just update those values. However I would prefer to use the ClientAPI if possible.

**Reply by tonygermano** [reply]
That would be the `/extensions/{extensionName}/properties` end point that he mentioned, but it's not really easy to use, because several of the fields contain serialized java objects.

https://localhost:8443/api/extensions/Data%20Pruner/properties
